### PR TITLE
Dataset histogram

### PR DESCRIPTION
**Pull request by igudich**
The simplest case (only for double and 2 dimensions including sparse).

### CODE REVIEW

File: core/include/scipp/core/dataset.h [hunk 1]
```diff
@@ -494,12 +494,20 @@ template <class Base> class MutableProxy : public Base {
 
 /// Const proxy for Dataset, implementing slicing and item selection.
 class SCIPP_CORE_EXPORT DatasetConstProxy {
+  explicit DatasetConstProxy() : m_dataset(nullptr) {}
```
Comment (SimonHeybrock) [inline core/include/scipp/core/dataset.h]: Why is this required?

File: core/include/scipp/core/dataset.h [hunk 2]
```diff
@@ -494,12 +494,20 @@ template <class Base> class MutableProxy : public Base {
...
+
 public:
   explicit DatasetConstProxy(const Dataset &dataset) : m_dataset(&dataset) {
     for (const auto &item : dataset.m_data)
       m_indices.emplace_back(item.first);
   }
 
+  static DatasetConstProxy makeShallowProxy(const Dataset &dataset) {
```
Comment (SimonHeybrock) [inline core/include/scipp/core/dataset.h]: The name is very unclear. How about making a more generic constructor that accepts a list of data items to keep:

```cpp
explicit DatasetConstProxy(const Dataset &dataset, const std::vector<std::string> &keys) : m_dataset(&dataset) {
    for (const auto &key : keys)
      m_indices.emplace_back(...);
  }
```

Then use it with an empty list of keys to get what `makeShallowProxy` does?

File: core/include/scipp/core/variable.h [hunk 3]
```diff
@@ -374,14 +374,15 @@ template <class T> Variable makeVariable(const Dimensions &dimensions) {
 }
 
 template <class T>
-Variable makeVariableWithVariances(const Dimensions &dimensions) {
+Variable makeVariableWithVariances(const Dimensions &dimensions,
+                                   units::Unit unit = units::dimensionless) {
   if (dimensions.sparse())
     return Variable(
         units::dimensionless, std::move(dimensions),
```
Comment (SimonHeybrock) [inline core/include/scipp/core/variable.h]: You forgot to change the unit here.

File: core/test/dataset_operations_test.cpp [hunk 4]
```diff
@@ -680,3 +681,34 @@ TYPED_TEST(DatasetBinaryOpTest, dataset_lhs_dataproxy_rhs) {
...
+}
+
+TEST(DatasetHistogram, simple_dataset_histogram) {
+  auto sparse = non_trivial_2d_sparse();
+  sparse.setSparseCoord(
+      "sparse1",
+      sparse["sparse"].coords()[sparse["sparse"].dims().sparseDim()]);
+  auto hist = core::histogram(
+      sparse, makeVariable<double>({Dim::Y, 6}, {1, 2, 3, 4, 5, 6}));
+  std::cout << hist << "\n";
```
Comment (SimonHeybrock) [inline core/test/dataset_operations_test.cpp]: Leftover cout from debugging?

File: core/test/dataset_operations_test.cpp [hunk 5]
```diff
@@ -680,3 +681,34 @@ TYPED_TEST(DatasetBinaryOpTest, dataset_lhs_dataproxy_rhs) {
     EXPECT_EQ(reference, item.data());
   }
 }
+
+Dataset non_trivial_2d_sparse() {
+  Dataset sparse;
+  auto var = makeVariable<double>({Dim::X, Dim::Y}, {3, Dimensions::Sparse});
+  var.sparseValues<double>()[0] = {1.5, 2.5, 3.5, 4.5, 5.5};
+  var.sparseValues<double>()[1] = {3.5, 4.5, 5.5, 6.5, 7.5};
+  var.sparseValues<double>()[2] = {1, 1, 2, 2, 2, 4, 4, 4, 6};
```
Comment (SimonHeybrock) [inline core/test/dataset_operations_test.cpp]: Can you add some data the is less than the first bin bound, to check if it dropped as it should be.

File: core/test/dataset_operations_test.cpp [hunk 6]
```diff
@@ -680,3 +681,34 @@ TYPED_TEST(DatasetBinaryOpTest, dataset_lhs_dataproxy_rhs) {
...
+                          hist.values<double>().end()};
+  for (scipp::index i = 0; i < static_cast<scipp::index>(res.size()); ++i)
+    EXPECT_EQ(ref[i], res[i]);
+}
+
+TEST(DatasetHistogram, simple_dataset_histogram) {
+  auto sparse = non_trivial_2d_sparse();
+  sparse.setSparseCoord(
+      "sparse1",
+      sparse["sparse"].coords()[sparse["sparse"].dims().sparseDim()]);
```
Comment (SimonHeybrock) [inline core/test/dataset_operations_test.cpp]: Maybe you could add `Dataset::setData` that accepts a `DataConstProxy` (currently it only exists for `Variable`). It would also copy the coord then. This would get rid of this mess that is currently required.

File: python/dataset.cpp [hunk 7]
```diff
@@ -134,6 +135,33 @@ void init_dataset(py::module &m) {
...
+        },
+        py::call_guard<py::gil_scoped_release>(),
+        "Returns a new Dataset with histograms for sparse dims");
+
+  m.def("histogram",
+        [](const Dataset &ds, const VariableConstProxy &bins) {
+          return core::histogram(ds, bins);
+        },
+        py::call_guard<py::gil_scoped_release>(),
+        "Returns a new Dataset with histograms for sparse dims");
```
Comment (SimonHeybrock) [inline python/dataset.cpp]: Can you make sure this is actually calling the overloads for `VariableConstProxy`? We have implicit conversion, so there is a risk that pybind11 is just calling the first overload, using costly conversion.

File: core/test/dataset_operations_test.cpp (hunk 6 above)
Comment (igudich) [reply core/test/dataset_operations_test.cpp]: Extracted to issue #373.

**Comment by SimonHeybrock** [comment]
Note that a generalized implementation will be supported once we solve #368.

We will probably need more testing in the future, but this is ok as a starting point.
